src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ImmutableCollectionCreator.cs

```diff
@@ -78,9 +79,11 @@ public override bool CreateImmutableDictionary(IDictionary items, out IDictionar
 
         private IEnumerable<KeyValuePair<string, TElement>> CreateGenericTElementIDictionary(IDictionary sourceDictionary)
         {
+#pragma warning disable 8605
             foreach (DictionaryEntry item in sourceDictionary)
+#pragma warning restore 8605
             {
-                yield return new KeyValuePair<string, TElement>((string)item.Key, (TElement)item.Value);
+                yield return new KeyValuePair<string, TElement>((string)item.Key, (TElement)item.Value!);
```

Comment:
Instead of disabling the warning:
```C#
private IEnumerable<KeyValuePair<string, TElement>> CreateGenericTElementIDictionary(IDictionary sourceDictionary)
{
    foreach (DictionaryEntry? item in sourceDictionary)
    {
        if (item.HasValue)
        {
            yield return new KeyValuePair<string, TElement>((string)item.Value.Key, (TElement)item.Value.Value!);
        }
    }
}
```